menoh/array.cpp

```diff
@@ -2,32 +2,47 @@
 
 namespace menoh_impl {
 
-    array::array(dtype_t d, std::vector<int> const& dims, void* data_handle)
+    array::array(dtype_t d, std::vector<int32_t> const& dims, void* data_handle)
```

Comment:
Why don't you use `uint32_t`? Is there any case that `dims` is negative?

Reply:
numpy accepts `-1` as dimension in some places (e.g. argument for `reshape`).
How about ONNX?
If ONNX also accepts `-1` in some places, then I think it makes sense to use signed integer type here too.

Reply:
If the element of `dims` takes `-1`, `int_32/64_t` makes sense.

Reply:
`Reshape` operator accepts `-1`. However `menoh::array` can't have -1 dim so It is reasonable to define two type of dims `std::vector<int64_t>` and `std::vector<uint64_t>` I think.
https://github.com/onnx/onnx/blob/master/docs/Operators.md#reshape

Reply:
Of course, I understand that `menoh::array()` and `menoh:array::dims()`  do not use `-1`.
But still I advocate that: if there are some places that uses dimension `-1` and signed vector types,  using the same signed vector type for here too is better for the consistency and simplicity of API.

Reply:
@okdshin @Hakuyume

Reply:
I choose `int32_t`. It is more simple.

Reply:
I like this idea https://github.com/pfnet-research/menoh/pull/41#discussion_r203271016. Users can know whether `-1` is accepted or not from the type of arguments.

Reply:
If using both `int32/uint32` looks confusing, it is ok to use `int32`. (In that case, I will wrap them using optional types in my own wrapper)